### PR TITLE
Create Jakarta EE 9 feature compatibility test

### PR DESCRIPTION
**Pull request by jhanders34**
- Create FAT test that validates that expected feature resolve with EE 9 features
- The FAT test also validates features that should not resolve with EE 9 features and validates they conflict on the eeCompatible feature
- Also the FAT test validates that EE7 and EE8 features will not start with EE9 features.  Some of the features have the same short name base, but different version number, but aren't singleton with each other because of the prefix change.  For example jsonp-1.0 and jsonp-2.0.  The test validates that they get conflicts instead of allowing them to start together.
- Update FeatureReplacementActions to have missing features in its list of EE features.
- Update features to have a dependency on eeCompatible so that the tests that should conflict, correctly conflict on the eeCompatible feature.
- Update scim-2.0 noship feature to work with Jakarta EE9.  The feature doesn't need any special transformation.  It just needs a tolerates on servlet-5.0.

### CODE REVIEW

**Comment by jhanders34** [comment]
#build

File: dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature [hunk 1]
```diff
@@ -6,6 +6,7 @@ IBM-ShortName: appClientSupport-1.0
 Subsystem-Name: Application Client Support for Server 1.0
 -features=com.ibm.websphere.appclient.appClient-1.0, \
   com.ibm.websphere.appserver.injection-1.0, \
+  com.ibm.websphere.appserver.eeCompatible-6.0; ibm.tolerates:="7.0, 8.0", \
```
Comment (tjwatson) [inline dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature:8]: Is there any thought to the placement of the eeCompatible feature?  Now that the feature manager actually pays attention to order of the included features (instead of using HashMap key order) perhaps it is best to put the eeCompatible inclusion first?  It makes the life of the resolver a bit easier because hard incompatibilities will be found earlier I think.

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+        // Remove test features that are in the FeatureReplacementActions
+        nonEE9JavaEEFeatures.remove("componenttest-1.0");
+        nonEE9JavaEEFeatures.remove("componenttest-2.0");
+        nonEE9JavaEEFeatures.remove("txtest-1.0");
+        nonEE9JavaEEFeatures.remove("txtest-2.0");
+        nonEE9JavaEEFeatures.remove("ejbTest-1.0");
+        nonEE9JavaEEFeatures.remove("ejbTest-2.0");
+
+        nonEE9MicroProfileFeatures.addAll(MicroProfileActions.MP10.getFeatures());
```
Comment (tjwatson) [inline dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:125]: the various MPXX FeatureSet objects would probably be better off as an enum that we can iterate over the values and the enum could indicate what EE versions it is associated with.  That way each time MP has a new release we don't have to update this code to add a new line.  Not really a blocker for this review.  Just an observation on where we could improve.

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+        // Add the MP convenience features as well
+        String[] mpVersions = { "1.0", "1.2", "1.3", "1.4", "2.0", "2.1", "2.2", "3.0",
+                                "3.2", "3.3", "4.0", "4.1" };
+        for (String mpVersion : mpVersions) {
+            nonEE9MicroProfileFeatures.add("microProfile-" + mpVersion);
+        }
+
+        // MP standalone features
+        nonEE9MicroProfileFeatures.add("mpContextPropagation-1.0");
```
Comment (tjwatson) [inline dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:146]: Is there an authoritative list of these standalone MP features?  Perhaps better maintained in MicroProfileActions?

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        if (feature.isDirectory() || !feature.getName().endsWith(".mf"))
+            return;
+
+        Scanner scanner = new Scanner(feature);
+        try {
+            String shortName = null;
+            while (scanner.hasNextLine()) {
+                String line = scanner.nextLine();
+                if (line.startsWith("IBM-ShortName:")) {
+                    shortName = line.substring("IBM-ShortName:".length()).trim();
```
Comment (tjwatson) [inline dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:200]: are we guaranteed to have IBM-ShortName header first?  Even for existing features that may not be using the `.feature` file?

**Comment by jhanders34** [comment]
#build

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+    @Test
+    public void testJsonP20Feature() throws Exception {
+        // jsonp-2.0 will conflict with itself
+        testCompatibility("jsonp-2.0", Collections.singletonMap("jsonp-2.0", "io.openliberty.jsonp"));
+    }
+
+    private void testCompatibility(String featureName, Map<String, String> specialConflicts) throws Exception {
+        final List<String> errors = new CopyOnWriteArrayList<>();
+        Result jakartaee91 = resolver.resolveFeatures(repository, Collections.<ProvisioningFeatureDefinition> emptySet(), Collections.singletonList("jakartaee-9.1"),
```
Comment (tjwatson) [inline dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java]: Seems this resolution is only here to get the set of kernel features.  Perhaps resolve something much more simple (e.g. osgiConsole-1.0).  But I'm not sure I understand how this code gets the kernel features though because nothing explicitly includes the kernel features that I know of so how would a resolve pull it in?

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java (hunk 4 above)
Comment (jhanders34) [reply dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:200]: We are scanning the entire file.  Not sure why you are thinking this is looking for IBM-ShortName to be first.  We check every line  unless we get to a line that tells us this feature is one we do not want to include.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java [hunk 6]
```diff
@@ -53,12 +53,15 @@
                                                  "mdb-3.2",
                                                  "componenttest-1.0",
                                                  "txtest-1.0",
-                                                 "websocket-1.1",
+                                                 "websocket-1.0",
```
Comment (Azquelt) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java]: Why this change? The `webProfile-7.0` feature includes websocket-1.1.

Comment (jhanders34) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/EE7FeatureReplacementAction.java]: Fair point.  I will add websocket-1.0 as a special case.

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java (hunk 5 above)
Comment (jhanders34) [reply dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java]: It gives me an empty list.  As such I am just removing this code.

File: dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature (hunk 1 above)
Comment (jhanders34) [reply dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature:8]: I had thought about that.  Are you thinking to do it for all features?

**Review by Azquelt — APPROVED** [review]

File: dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature (hunk 1 above)
Comment (tjwatson) [reply dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature:8]: There is a risk to reordering all features because it may influence the first choice made by the resolver for a solution based on the first preferred version found.  So that could change the features that get enabled for a specific configuration when multiple valid solutions exist.  Here I was really just questioning the addition of the eeCompatible and if we should put that first here.

Comment (jhanders34) [reply dev/com.ibm.websphere.appserver.features/visibility/public/appClientSupport-1.0/com.ibm.websphere.appserver.appClientSupport-1.0.feature:8]: I updated it to be first.  I made sure it was before other tolerate features originally, but it is fine to be first.

**Comment by jhanders34** [comment]
#libby
#build

File: dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java (hunk 2 above)
Comment (tevans78) [reply dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java:125]: MicroProfileActions.ALL is a `Set<FeatureSet>` that contains all versions

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#build

**Comment by jhanders34** [comment]
#libby #build

**Review by tevans78 — APPROVED** [review]

**Comment by jhanders34** [comment]
#libby
